stew/conio.nim

```diff
@@ -0,0 +1,337 @@
+       importc: "WriteConsoleW", stdcall, dynlib: "kernel32", sideEffect.}
+  proc writeFile(hFile: uint, lpBuffer: pointer,
+                 nNumberOfBytesToWrite: uint32,
+                 lpNumberOfBytesWritten: var uint32,
+                 lpOverlapped: pointer): int32 {.
+       importc: "WriteFile", dynlib: "kernel32", stdcall, sideEffect.}
+  proc getStdHandle(nStdHandle: uint32): uint {.
+       importc: "GetStdHandle", stdcall, dynlib: "kernel32", sideEffect.}
+  proc wideCharToMultiByte(codePage: cuint, dwFlags: uint32,
+                           lpWideCharStr: ptr Utf16Char, cchWideChar: cint,
+                           lpMultiByteStr: ptr char, cbMultiByte: cint,
+                           lpDefaultChar: pointer,
+                           lpUsedDefaultChar: pointer): cint {.
+       importc: "WideCharToMultiByte", stdcall, dynlib: "kernel32", sideEffect.}
+  proc getFileType(hFile: uint): uint32 {.
+       importc: "GetFileType", stdcall, dynlib: "kernel32", sideEffect.}
+
+  const
+    CP_UTF8 = 65001'u32
+    STD_INPUT_HANDLE = cast[uint32](-10)
+    STD_OUTPUT_HANDLE = cast[uint32](-11)
+    INVALID_HANDLE_VALUE = cast[uint](-1)
+    ENABLE_PROCESSED_INPUT = 0x0001'u32
+    ENABLE_ECHO_INPUT = 0x0004'u32
+    FILE_TYPE_CHAR = 0x0002'u32
+
+  proc isConsoleRedirected(hConsole: uint): bool =
+    ## Returns ``true`` if console handle was redirected.
+    let res = getFileType(hConsole)
+    if res == FILE_TYPE_CHAR:
+      # The specified handle is a character device, typically an LPT device or a
+      # console.
+      false
+    else:
+      true
+
+  proc readConsoleInput(maxBytes: int): IoResult[string] =
+    let hConsoleInput =
+      block:
+        let res = getStdHandle(STD_INPUT_HANDLE)
+        if res == INVALID_HANDLE_VALUE:
+          return err(ioLastError())
+        res
+
+    let prevInputCP =
+      block:
+        let res = getConsoleCP()
+        if res == cuint(0):
+          return err(ioLastError())
+        res
+
+    if isConsoleRedirected(hConsoleInput):
+      # Console STDIN is redirected, we should use ReadFile(), because
+      # ReadConsole() is not working for such types of STDIN.
+      if setConsoleCP(CP_UTF8) == 0'i32:
+        return err(ioLastError())
+
+      # Allocating buffer with size equal to `maxBytes` + len(CRLF)
+      var buffer = newString(maxBytes + 2)
+      let bytesToRead = uint32(len(buffer))
+      var bytesRead: uint32
+      let rres = readFile(hConsoleInput, cast[pointer](addr buffer[0]),
+                          bytesToRead, bytesRead, nil)
+      if rres == 0:
+        let errCode = ioLastError()
+        discard setConsoleCP(prevInputCP)
+        return err(errCode)
+
+      if setConsoleCP(prevInputCP) == 0'i32:
+        return err(ioLastError())
+
+      # Truncate additional bytes from buffer.
+      buffer.setLen(int(min(bytesRead, uint32(maxBytes))))
+
+      # Trim CR/CRLF from buffer.
+      if len(buffer) > 0:
```

Comment:
This looks like duplicating `strutils.stripLineEnd`.

Reply:
The problem of `strutils.stripLineEnd` is that it thinks that `\v` and `\f` are line end, but this is not a proper markers of end of line.

Reply:
These characters are quite obscure nowadays, but OK - if this is a general purpose procedure for reading a line of characters from the screen, why should it preserve a page break character at the end of the line, but not a new line character in the same place?

Reply:
@zah we are supposed to read UTF-8 encoded strings are you sure `\v` and `\f` could not be part of UTF-8 encoded symbol which was cut because of buffer limits?

Reply:
Yes, in UTF-8, the first 128 characters in the ASCII table are encoded with a single byte that has a leading zero bit. All other characters are encoded with a sequence of bytes where each byte has a leading bit set to 1. More details here:

https://en.wikipedia.org/wiki/UTF-8#Encoding

Reply:
@zah  you are right, but i still think `\v` and `\f` is not a proper condition for line end which we are going to receive from console. When we reading from console the only proper line end is `\n` and `\r\n` and nothing else, `strutils` procedure is incorrect.